### DISCUSSION TITLE
Wide angle model tracing

### DISCUSSION THREAD

**Discussion by GY1015** [mjhoptics/ray-optics#157] [Q&A] [answered]
Hi!
I'm trying learn rayoptics and trying to open some codev samples.
I encountered an issue while running the sample lens file "fisheye.seq" in Code V. The sd value seems to be fixed at 1, and the resulting image is rendered abnormally. I tested the configuration by gradually changing the angle, and everything worked fine up to 50 degrees. However, at 60 degrees, I received a "tuple index out of range" error. I'll also attach images for your reference.
![image](https://github.com/user-attachments/assets/8aa8b195-9b53-4b85-9840-492a9cab845d)
![image](https://github.com/user-attachments/assets/6649806f-5725-48f0-b150-a5b889606d35)

I'm wondering if ray tracing for wide-angle lenses is not fully supported in RayOptics, or if I might be missing something in the setup. Do you have any suggestions or potential solutions to this issue?

I also wonder it is possible to set surface type as odd polynomial or Qcon Asphere.

**Comment by mjhoptics** [comment]
Hello @GY1015,
I don't have access to the model you're using so I can't diagnose the problem directly. If you post the results of:
```
listobj(sm)
listobj(osp)
```
I could take a look.

My expectation is the ray trace *should* work up to 90 deg. Beyond that, the ray aiming scheme would need changes, I'd wager. Modeling wide angle systems is challenging and it wouldn't surprise me if this system  has some combination of attributes that defeats the current ray trace scheme.
Regards,
Mike Hayford

**Reply by GY1015** [reply]
Is the error I mentioned(tuple index out of range) also occurring because the angle exceeds 90 degrees?

This is the results of 
listobj(sm)
listobj(osp)


0: Obj: dummy
profile: Spherical
c=0.0,   r=0.0
surface_od=1143005230286.0862
t=100000000000.0 (+1);    medium: air

1: transmit
profile: Spherical
c=0.20944601528955914,   r=4.7745
surface_od=2.5052022780735457
t=0.1299 (+1);    medium: 713.539

2: transmit
profile: Spherical
c=1.1336583153837434,   r=0.8821
surface_od=1.0
t=0.5688 (+1);    medium: air

3: transmit
profile: Spherical
c=0.5944596361907026,   r=1.6822000000000001
surface_od=1.0
t=0.0938 (+1);    medium: 618.633

4: transmit
profile: Spherical
c=1.2400793650793651,   r=0.8064
surface_od=1.0
t=0.3313 (+1);    medium: air

5: transmit
profile: Spherical
c=0.6148170919151552,   r=1.6265
surface_od=1.0
t=0.4838 (+1);    medium: 595.392

6: transmit
profile: Spherical
c=-1.372872048325096,   r=-0.7284
surface_od=1.0
t=0.0969 (+1);    medium: 623.569

7: transmit
profile: Spherical
c=0.12000336009408263,   r=8.3331
surface_od=1.0
t=0.05 (+1);    medium: air

8: transmit
profile: Spherical
c=0.13100322267927791,   r=7.6334
surface_od=1.0
t=0.1875 (+1);    medium: 805.409

9: transmit
profile: Spherical
c=0.0,   r=0.0
surface_od=1.0
t=0.0806 (+1);    medium: air

10: transmit
profile: Spherical
c=0.0,   r=0.0
surface_od=1.0
t=0.0938 (+1);    medium: 581.410

11: transmit
profile: Spherical
c=0.0,   r=0.0
surface_od=1.0
t=0.1413 (+1);    medium: air

12 (stop): transmit
profile: Spherical
c=0.0,   r=0.0
surface_od=1.0
t=0.1 (+1);    medium: air

13: transmit
profile: Spherical
c=-0.34814092744743075,   r=-2.8724
surface_od=1.0
t=0.05 (+1);    medium: 593.352

14: transmit
profile: Spherical
c=0.35670970963829635,   r=2.8034
surface_od=1.0
t=0.25 (+1);    medium: 640.600

15: transmit
profile: Spherical
c=-0.8685833405715279,   r=-1.1513
surface_od=1.0
t=0.0063 (+1);    medium: air

16: transmit
profile: Spherical
c=0.30690851057299817,   r=3.2583
surface_od=1.0
t=0.2919 (+1);    medium: 488.701

17: transmit
profile: Spherical
c=-1.1328877308258751,   r=-0.8827
surface_od=1.0
t=0.0913 (+1);    medium: 805.254

18: transmit
profile: Spherical
c=-0.5842486562280906,   r=-1.7116
surface_od=1.0
t=2.30568559121 (+1);    medium: air

19: Img: dummy
profile: Spherical
c=0.0,   r=0.0
surface_od=1.5
radius=1.5

do apertures: True
aperture: object pupil; value=0.25
field: object angle; value=0.0
x,y=0.0 vlx=-0.000 vux=-0.000 vly=-0.000 vuy=-0.000
y=45.0 vly=-0.204 vuy=-0.276 vlx=-0.065 vux=-0.065
y=85.0 vly=-4.358 vuy=-4.898 vlx=-0.255 vux=-0.255
central wavelength=550.0 nm
wavelength (weight) =  620.0000 (1.000),  550.0000 (1.000)*,  480.0000 (1.000)
focus shift=-0.00288898390382

**Reply by mjhoptics** [reply, marked as answer]
Quick update: there are several problems affecting this model:

1. ray failures are not handled gracefully when tracing the rays thru the edges of the pupil. I have a fix in work for this.
2. fisheye lenses have large pupil aberration. Rays close to 90 deg go through an entrance pupil 50% closer to the front vertex than the paraxial pupil location. ray-optics doesn't handle this well and it will take some time to work out an automated solution.
3. I see some small changes to make manual specification of the chief ray aim point for wide angle systems possible. It'll probably be bundled with point 1.
4. the model also uncovered that object tangents rather than angles were being used to calculate fractional object points. This will be fixed.

When I update the short term fixes above, I'll post a notebook with how to use ray-optics to trace wide-angle lenses.
Feel free to ask questions.
Regards
Mike Hayford

**Reply by GY1015** [reply]
Thank you Mike! Can you tell me when the update will be?

**Reply by mjhoptics** [reply]
Hello @GY1015,
This is turning out to be more involved than I thought. I think it should be possible to trace out to a 90 deg field angle with some effort. Going beyond 90 deg will be more challenging. I'm on travel this week and having written a lot of software, I wouldn't plan on having an update for several weeks.
Sorry for the delay but thanks for the interesting problem.
Regards
Mike Hayford

**Comment by mjhoptics** [comment]
Hello @GY1015,

It's taken a while, but I think **ray-optics** is much more likely to trace a fisheye or similar lens using v0.9.5. Below is some background about the source of the issues and how it was solved. The only input change for your original model would be to add the kwarg `is_wide_angle=True` to the FieldSpec creation below.

The original example was a fisheye lens from the CODE V example library. The then current version of **ray-optics** could trace that lens out to about 70 deg. Investigation led to the use of a separate `is_wide_angle` mode in `osp['fov']`. Setting this to True invokes a separate set of calculations that better model the behavior of the entrance pupil, especially for 90 deg and greater fields of view.

While investigating different models, it became apparent that being able to specify a field point in terms of the real ray height in the image plane would substantially increase the number of models that were accessible to ray-optics. This can be done now using the 'real height' field value type in image space.

The [wideangle module](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.raytr.wideangle.html#module-rayoptics.raytr.wideangle) is where most of the new capability has implemented. It was introduced in v0.9.3 and refined in v0.9.5. The goal is for the user *not* to have to know about this module, but several functions in the module are used to calculate spherical aberration of the pupil in the notebook below.

[git-hub discussion](https://github.com/mjhoptics/ray-optics/discussions/157)


```python
%matplotlib inline
```


```python
isdark = False
```


```python
from scipy import interpolate

from rayoptics.environment import *
from rayoptics.raytr.wideangle import *
```

# Wide Angle model

## Use new setting in FieldSpec, is_wide_angle

The is_wide_angle=True mode in FieldSpec uses a completely different way of aiming rays to go through the entrance pupil. A distinguishing feature of wide angle or fisheye lenses is that the location of the entrance pupil can change dramatically from the paraxial position. Aiming rays at the paraxial entrance pupil plane becomes less effective as the field angle increases until it fails completely at 90 degrees.

When is_wide_angle=True, a search is conducted parameterized by z_enp, the distance along the z axis, measured from the first surface, to the real entrance pupil for that field angle. At extreme fields, ray failures during the search require a two phase search approach: the first is to find a range of z_enp values that trace to the actual stop surface without ray failure. Once this range is determined, a standard ray iteration procedure is applied to find the exact z_enp going through the center of the stop surface.

Once the chief ray has been determined for a field angle, the entrance pupil is taken as a disk perpendicular to the chief ray, rather than the optical axis. This approximates the actual behavior of the entrance pupil in wide angle lenses and avoids the foreshortening that occurs when using a plane perpendicular to the optical axis.


```python
opm = OpticalModel()
sm  = opm['seq_model']
osp = opm['optical_spec']
fov = osp['fov']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
osp['pupil'] = PupilSpec(osp, key=['object', 'epd'], value=0.25)
fov = osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=90., flds=[0., 0.7071, 1.0], 
                             is_relative=True, is_wide_angle=True)
osp['wvls'] = WvlSpec([(480., 1.0), (550, 1.0), (620., 1.0)], ref_wl=1)

obj_dist = 1.0e10
sm.gaps[0].thi = obj_dist

sm.do_apertures = False
opm.radius_mode = True

sm.add_surface([4.7745, .1299, 1.713, 53.9], sd=2.0)
sm.add_surface([0.8821, .5688, 'air'], sd=0.88)
sm.add_surface([1.6822, .0938, 1.618, 63.3], sd=0.95)
sm.add_surface([0.8064, .3313, 'air'], sd=0.725)
sm.add_surface([1.6265, .4838, 1.595, 39.2], sd=0.6)
sm.add_surface([-0.7284, .0938, 1.623, 56.9], sd=0.6)
sm.add_surface([8.3331, .05, 'air'], sd=0.6)
sm.add_surface([7.6334, .1875, 1.805, 40.9], sd=0.6)
sm.add_surface([0, .0806, 'air'], sd=0.6)
sm.add_surface([0, .0938, 1.581, 41.0], sd=0.6)
sm.add_surface([0, .1413, 'air'], sd=0.6)
sm.add_surface([0, .1000, 'air'], sd=0.284)
sm.set_stop()
sm.add_surface([-2.8724, .0500, 1.713, 53.2], sd=0.6)
sm.add_surface([2.8034, .2500, 1.640, 60.0], sd=0.6)
sm.add_surface([-1.1513, .0063, 'air'], sd=0.6)
sm.add_surface([3.2583, .2919, 1.488, 70.1], sd=0.6)
sm.add_surface([-0.8827, .0913, 1.805, 25.4], sd=0.6)
sm.add_surface([-1.7116, 2.305685559121, 'air'], sd=0.6)

opm.update_model()
```


```python
sm.gaps[-1].thi = pm.thi_ht_solve(pm.ax, -2, 0.)
opm.update_model()
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+10       air             1      1.0000
        1:     4.774500     0.129900   713.539             1      2.0000
        2:     0.882100     0.568800       air             1     0.88000
        3:     1.682200    0.0938000   618.633             1     0.95000
        4:     0.806400     0.331300       air             1     0.72500
        5:     1.626500     0.483800   595.392             1     0.60000
        6:    -0.728400    0.0938000   623.569             1     0.60000
        7:     8.333100    0.0500000       air             1     0.60000
        8:     7.633400     0.187500   805.409             1     0.60000
        9:     0.000000    0.0806000       air             1     0.60000
       10:     0.000000    0.0938000   581.410             1     0.60000
       11:     0.000000     0.141300       air             1     0.60000
     Stop:     0.000000     0.100000       air             1     0.28400
       13:    -2.872400    0.0500000   713.532             1     0.60000
       14:     2.803400     0.250000   640.600             1     0.60000
       15:    -1.151300   0.00630000       air             1     0.60000
       16:     3.258300     0.291900   488.701             1     0.60000
       17:    -0.882700    0.0913000   805.254             1     0.60000
       18:    -1.711600      2.84061       air             1     0.60000
      Img:     0.000000      0.00000                       1      1.0000


## Set vignetting

Similar to the difficulties finding the chief ray at large field angles, determining the vignetting factors for large fields is also troubled by ray failures. A new option, use_bisection, was added to the set_vignetting function and is True by default if the model is_wide_angle=True. use_bisection will use a bisection algorithm to find the limiting apertures for the edge rays at each field.


```python
cmds.set_vignetting(opm)
```


```python
listobj(fov)
```

    field: object angle; value=90.0
    is_relative=True, is_wide_angle=True
    x,y=0.0 vlx= 0.001 vux= 0.001 vly= 0.001 vuy= 0.001
    y=0.7071 vly= 0.286 vuy= 0.229 vlx=-0.134 vux=-0.134
    y=1.0 vly= 0.659 vuy= 0.604 vlx=-0.288 vux=-0.288
    



```python
layout_pltw = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=False).plot()
```

![output_12_0](https://github.com/user-attachments/assets/e994ae5b-a98c-4409-9965-740f5ecdf161)


## Calculate the position of the entrance pupil, z_enp, across the FOV

The function [find_real_enp](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.raytr.wideangle.html#rayoptics.raytr.wideangle.find_real_enp) may be used to find z_enp at a field point. The following cells loop over the entire field of view and prints and plots the results. The entrance pupil moves about 75% of the distance from the paraxial pupil to the front surface (about 0.66 inches) at the 90 deg field angle. 


```python
z_enps = []
flds = []
obj_angs = []
cwl = osp['wvls'].central_wvl
print("       py      obj angle     z_enp")
for i, fld_ht in enumerate(np.arange(-1., 1.1, .1)):
    fld = Field(y=fld_ht)
    z_enp, rr = find_real_enp(opm, sm.stop_surface, fld, cwl)
    d0 = rr.pkg.ray[0][mc.d]
    ang_x = np.rad2deg(math.atan2(d0[0], d0[2]))
    ang_y = np.rad2deg(math.atan2(d0[1], d0[2]))
    print(f"{i:2d}   {fld_ht:6.3f}:    {ang_y:7.2f}    {z_enp:8.4f}")
    flds.append(fld_ht)
    obj_angs.append(ang_y)
    z_enps.append(z_enp)

tck = interpolate.splrep(obj_angs, z_enps)

polyline = np.linspace(obj_angs[0], obj_angs[-1], 50)
z_enps_intrp = interpolate.splev(polyline, tck, der=0)
```

           py      obj angle     z_enp
     0   -1.000:     -90.00      0.2075
     1   -0.900:     -81.00      0.3743
     2   -0.800:     -72.00      0.5064
     3   -0.700:     -63.00      0.6098
     4   -0.600:     -54.00      0.6899
     5   -0.500:     -45.00      0.7512
     6   -0.400:     -36.00      0.7971
     7   -0.300:     -27.00      0.8304
     8   -0.200:     -18.00      0.8529
     9   -0.100:      -9.00      0.8659
    10   -0.000:      -0.00      0.8702
    11    0.100:       9.00      0.8659
    12    0.200:      18.00      0.8529
    13    0.300:      27.00      0.8304
    14    0.400:      36.00      0.7971
    15    0.500:      45.00      0.7512
    16    0.600:      54.00      0.6899
    17    0.700:      63.00      0.6098
    18    0.800:      72.00      0.5064
    19    0.900:      81.00      0.3743
    20    1.000:      90.00      0.2075



```python
plt.scatter(z_enps, obj_angs)
plt.plot(z_enps_intrp, polyline)
plt.title("Pupil Aberration")
plt.ylabel("object angle")
plt.xlabel("z_enp")
plt.show()
```

![output_15_0](https://github.com/user-attachments/assets/2c2ca781-12ad-46c8-8f82-8be78d0c6214)


# Switch to real image height specification

It is often both desireable and convenient to specify field via (paraxial) height in the image plane. For wide angle systems, this is problematic because the paraxial and real image heights can be very different due to the enormous image distortion in fish-eye lenses. What is needed is a specification in terms of the real image height. A new field value type has been added for image space specification: 'real height'.

## Get the image height of the chief ray at full field


```python
cr_max, cr_max_exp_seg = fov.fields[-1].chief_ray
list_ray(cr_max)
```

                X            Y            Z           L            M            N               Len
      0:      0.00000 -10000000000.20749        1e+10     0.000000     1.000000     0.000000        1e+10
      1:      0.00000     -1.39221      0.20749     0.000000     0.775332     0.631554      0.74323
      2:      0.00000     -0.81596      0.54698     0.000000     0.613892     0.789390      0.21258
      3:      0.00000     -0.68546      0.14599     0.000000     0.537721     0.843123      0.21672
      4:      0.00000     -0.56893      0.23491     0.000000     0.416447     0.909160      0.18919
      5:      0.00000     -0.49014     0.075609     0.000000     0.373901     0.927469      0.33431
      6:      0.00000     -0.36514    -0.098133     0.000000     0.354310     0.935128      0.21066
      7:      0.00000     -0.29050    0.0050652     0.000000     0.551934     0.833888     0.059111
      8:      0.00000     -0.25788    0.0043572     0.000000     0.321481     0.946916      0.19341
      9:      0.00000     -0.19570            0     0.000000     0.581659     0.813433     0.099086
     10:      0.00000     -0.13807            0     0.000000     0.367198     0.930143      0.10084
     11:      0.00000     -0.10104            0     0.000000     0.581659     0.813433      0.17371
     12:      0.00000      0.00000            0     0.000000     0.581659     0.813433      0.12186
     13:      0.00000      0.07088  -0.00087475     0.000000     0.350405     0.936598     0.055877
     14:      0.00000      0.09046      0.00146     0.000000     0.367686     0.929950      0.25153
     15:      0.00000      0.18295    -0.014629     0.000000     0.497631     0.867389       0.0311
     16:      0.00000      0.19842    0.0060474     0.000000     0.311677     0.950188      0.25373
     17:      0.00000      0.27751    -0.044757     0.000000     0.312159     0.950030      0.11288
     18:      0.00000      0.31274    -0.028815     0.000000     0.414816     0.909905       3.1535
     19:      0.00000      1.62089            0     0.000000     0.414816     0.909905            0


## Change the field specification to 'real height' in image space.


```python
listobj(fov.set_key_value(('image', 'real height'), cr_max.ray[-1][mc.p][1]))
```

    field: image real height; value=1.6208874188590325
    is_relative=True, is_wide_angle=True
    x,y=0.0 vlx= 0.001 vux= 0.001 vly= 0.001 vuy= 0.001
    y=0.7071 vly= 0.286 vuy= 0.229 vlx=-0.134 vux=-0.134
    y=1.0 vly= 0.659 vuy= 0.604 vlx=-0.288 vux=-0.288
    


## Evaluate pupil spherical aberration

To evaluate pupil spherical for real image height specifications, use the [eval_real_image_ht](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.raytr.wideangle.html#rayoptics.raytr.wideangle.eval_real_image_ht) to calculate z_enp across the field of view.


```python
z_enps = []
flds = []
obj_angs = []
cwl = osp['wvls'].central_wvl
print("       py      obj angle     z_enp")
for i, fld_ht in enumerate(np.arange(-1., 1.1, .1)):
    fld = Field(y=fld_ht)
    (p0, d0), z_enp = eval_real_image_ht(opm, fld, cwl)
    ang_x = np.rad2deg(math.atan2(d0[0], d0[2]))
    ang_y = np.rad2deg(math.atan2(d0[1], d0[2]))
    print(f"{i:2d}   {fld_ht:6.3f}:    {ang_y:7.2f}    {z_enp:8.4f}")
    flds.append(fld_ht)
    obj_angs.append(ang_y)
    z_enps.append(z_enp)

tck = interpolate.splrep(obj_angs, z_enps)

polyline = np.linspace(obj_angs[0], obj_angs[-1], 50)
z_enps_intrp = interpolate.splev(polyline, tck, der=0)
```

           py      obj angle     z_enp
     0   -1.000:     -90.00      0.2075
     1   -0.900:     -76.81      0.4397
     2   -0.800:     -66.32      0.5746
     3   -0.700:     -56.95      0.6659
     4   -0.600:     -48.19      0.7314
     5   -0.500:     -39.79      0.7795
     6   -0.400:     -31.63      0.8148
     7   -0.300:     -23.61      0.8401
     8   -0.200:     -15.70      0.8571
     9   -0.100:      -7.84      0.8670
    10   -0.000:      -0.00      0.8702
    11    0.100:       7.84      0.8670
    12    0.200:      15.70      0.8571
    13    0.300:      23.61      0.8401
    14    0.400:      31.63      0.8148
    15    0.500:      39.79      0.7795
    16    0.600:      48.19      0.7314
    17    0.700:      56.95      0.6659
    18    0.800:      66.32      0.5746
    19    0.900:      76.81      0.4397
    20    1.000:      90.00      0.2075



```python
plt.scatter(z_enps, obj_angs)
plt.plot(z_enps_intrp, polyline)
plt.title("Pupil Aberration")
plt.ylabel("object angle")
plt.xlabel("z_enp")
plt.show()
```

   
![output_23_0](https://github.com/user-attachments/assets/732d9f31-9684-430b-b974-b092577c64b2)
    

# 270 deg fisheye

## Model from the OpticalBenchHub of the PhotonsToPhotos website.

Many of the wide angle models on this website have real image height specifications.


```python
lens_url = "https://www.photonstophotos.net/GeneralTopics/Lenses/OpticalBench/OpticalBench.htm#Data/US003524697_Example02.txt"

opm, import_info = open_model(lens_url, info=True)
sm  = opm['seq_model']
osp = opm['optical_spec']
fov = osp['fov']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```


```python
layout_pltwb = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, do_draw_beams=False, do_draw_ray_fans=False, is_dark=False).plot()
```

    
![output_26_0](https://github.com/user-attachments/assets/a66f5102-c74f-4108-9d3e-45c79bc61d01)


## Evaluate pupil spherical aberration

To evaluate pupil spherical for real image height specifications, use the [eval_real_image_ht](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.raytr.wideangle.html#rayoptics.raytr.wideangle.eval_real_image_ht) to calculate z_enp across the field of view.

The field of view of this lens is so wide that the real entrance pupil is outside of the first surface vertex for angles larger than 225 deg total field of view!


```python
z_enps = []
flds = []
obj_angs = []
cwl = osp['wvls'].central_wvl
print("       py      obj angle     z_enp")
for i, fld_ht in enumerate(np.arange(-1., 1.1, .1)):
    fld = Field(y=fld_ht)
    (p0, d0), z_enp = eval_real_image_ht(opm, fld, cwl)
    ang_x = np.rad2deg(math.atan2(d0[0], d0[2]))
    ang_y = np.rad2deg(math.atan2(d0[1], d0[2]))
    print(f"{i:2d}   {fld_ht:6.3f}:    {ang_y:7.2f}    {z_enp:8.4f}")
    flds.append(fld_ht)
    obj_angs.append(ang_y)
    z_enps.append(z_enp)

tck = interpolate.splrep(obj_angs, z_enps)

polyline = np.linspace(obj_angs[0], obj_angs[-1], 50)
z_enps_intrp = interpolate.splev(polyline, tck, der=0)
```

           py      obj angle     z_enp
     0   -1.000:    -134.31    -46.3155
     1   -0.900:    -119.05     -8.9098
     2   -0.800:    -104.61     11.1892
     3   -0.700:     -90.55     23.5221
     4   -0.600:     -76.81     31.5258
     5   -0.500:     -63.41     36.8702
     6   -0.400:     -50.32     40.4781
     7   -0.300:     -37.49     42.8904
     8   -0.200:     -24.87     44.4306
     9   -0.100:     -12.40     45.2886
    10   -0.000:      -0.00     45.5643
    11    0.100:      12.40     45.2886
    12    0.200:      24.87     44.4306
    13    0.300:      37.49     42.8904
    14    0.400:      50.32     40.4781
    15    0.500:      63.41     36.8702
    16    0.600:      76.81     31.5258
    17    0.700:      90.55     23.5221
    18    0.800:     104.61     11.1892
    19    0.900:     119.05     -8.9098
    20    1.000:     134.31    -46.3155



```python
plt.scatter(z_enps, obj_angs)
plt.plot(z_enps_intrp, polyline)
plt.title("Pupil Aberration")
plt.ylabel("object angle")
plt.xlabel("z_enp")
plt.show()
```


 ![output_29_0](https://github.com/user-attachments/assets/24ad9533-df01-4fdc-9ef1-469ffe3ad28c)
